### DISCUSSION TITLE
Query filter for `not null`

### DISCUSSION THREAD

**Discussion by kennethacohen** [supabase/supabase#1133] [Questions] [answered]
This is incredibly basic (apologies), but for some reason I can't figure out how to compose a query that gets non-null values in a column

```ts
const { data: topRecipes } = await db
    .from<IRecipeSearch>("recipe_searches")
    .select("*")
    .not("external_recipe_id", "eq", null)

--

const { data: topRecipes } = await db
    .from<IRecipeSearch>("recipe_searches")
    .select("*")
    .neq("external_recipe_id", null)
```

Neither of the above seem to be working - would appreciate any help here! 

Related: it would be nice to debug and see the query that SB is generating under the hood

**Comment by kennethacohen** [comment]
Actually the `neq` method above worked for me in a previous project...but doesn't seem to work now, so this may just be a bug?

**Comment by steve-chavez** [comment, marked as answer]
Check the [is()](https://supabase.io/docs/reference/javascript/is#with-select) filter.

**Reply by kiwicopple** [reply]
Should be able to be combined with the `not` filter too:

```js
const { data: topRecipes } = await db
    .from<IRecipeSearch>("recipe_searches")
    .select("*")
    .not("external_recipe_id", "is", null)
```

**Reply by kennethacohen** [reply]
Using `.is("external_recipe_id", false)` doesn't work, but @kiwicopple usage of `not()` works, thanks folks!

**Reply by chrskerr** [reply]
I'm struggling with the not() pattern here using the below code, the target field is a nullable timestamptz. Never returns any rows despite me not having added a deleted_at value to any row.

Have there been any changes to the pattern here by any chance? Thanks for the advice!

`const { data, error } = await supabase.from( "offers" ).select( "*" ).eq( "user", authUser.id ).not( "deleted_at", "is", null );`

**Reply by steve-chavez** [reply]
@chrskerr Could you try with `"null"`(inside double quotes)?. Like:

```js
const { data, error } = await supabase.from( "offers" )
.select( "*" )
.eq( "user", authUser.id )
.not( "deleted_at", "is", "null" );
```

**Reply by chrskerr** [reply]
Thanks @steve-chavez - just tried `const { data } = await supabase.from( "offers" ).select( "*" ).eq( "user", authUser.id ).not( "deleted_at", "is", "null" );` and it still returns `[]`

**Reply by steve-chavez** [reply]
@chrskerr I've just tried the `not("col", "is", ..)` method locally, and it works for me both with `null` and `"null"`.

> const { data } = await supabase.from( "offers" ).select( "*" ).eq( "user", authUser.id ).not( "deleted_at", "is", "null" ); and it still returns []

Perhaps it's just that your `eq` filter contains a user whose `deleted_at` is null? Maybe try removing the `eq` filter for your tests.

**Reply by awaaisali** [reply]
Also not working for me.
**const {data, error} = await supabase
    .from("properties")
    .select("*")
    .not("properties_id","is", null);**

**Reply by anirudh5harma** [reply]
Same issue! Doesn't work for me either

@steve-chavez can you please assist with this.

**Reply by steve-chavez** [reply]
Hm, @soedirgo was there something changed with `postgrest-js` [is](https://supabase.com/docs/reference/javascript/is)?

**Reply by soedirgo** [reply]
I don't think so, this should work as before. It'd help if you provide a reproducible example demonstrating the issue, otherwise it's hard to debug this.

**Reply by qqpann** [reply]
I tried it but doesn't work with TypeScript

**Reply by danya0365** [reply]
Test and worked

```
    if (isClaim !== undefined) {
      if (!isClaim) {
        query = query.is('claim_at', null);
      } else {
        query = query.not('claim_at', 'is', null);
      }
    }
 ```

**Comment by maheshj01** [comment]
Is it possible to have a non empty value in supabase database. I see that by default an empty cell is considered as "null" ([null] in database) which is treated as null in dart. So I was wondering if there is a way to consider an empty cell as "" (empty string) by default when we create the column?

**Reply by cdimitroulas** [reply]
You should set the column's default value to be empty string in that case.

**Comment by mike-wax** [comment]
If you're doing this with a JSONB column, make sure you use `->>` (as opposed to `->`) on the final key in the JSON

```
const result = await supabase
      .from('apps')
      .select()
      .not('details->schedule->>interval', 'is', null)
```

**Comment by takenoto** [comment]
I'm facing the same problem today (2023-07-07).
Platform: Flutter/Dart

Query:
`supabase
    .from('vehicles_positions)
    .select('vehicle_id, latitude, longitude)
    .in_('vehicle_id', ids)
    .not('latitude','is',null)
`

I'm trying to get the vehicles where the latitude is not null.
Already tried:
.not('latitude','is',null)

.not('latitude','is',"null")

.not('latitude','is',"")

.neq('latitude',null)

.neq('latitude', 'null')

.neq('latitude', "")

I also tried variations such as 'NULL' and 'Null'.

**Reply by earonesty** [reply]
ive just changed all my "nullable" columns to default to empty string to work around this issue

**Comment by krystian-dev548** [comment]
Hi to all, is there any other way to archive "is not null" without using "default to empty string"?

**Reply by GaryAustin1** [reply]
.not('col','is', null)

Verified it is working in javascript client.

**Reply by krystian-dev548** [reply]
Hi, many thanks. I confirm that works also on Flutter

**Reply by flogy** [reply]
Working for me as well, thank you!

**Reply by krystian-dev548** [reply]
Hi, thanks. It works also on Flutter as well

**Comment by AndrewJudson** [comment]
What is the idiomatic way to do this in the Python client?

**Reply by Yumat10** [reply]
Did you ever find an answer for this?

**Reply by GravO8** [reply]
I think it is `.not_.is_("columnName", "null")`
[Source](https://supabase.com/docs/reference/python/not)

**Comment by AVivero** [comment]
It worked for me using `query.not("columnName", "is", "NULL")` and `query.is("columnName", "NULL")`. Hope that helps somebody!

**Comment by pmnzt** [comment]
while yes ```.not('col','is', null)``` works, but what if we need the 'is not null' in a .or ?
like ```.or('col.is_not.null')```

**Reply by GaryAustin1** [reply]
The PostgREST documentation is the best source  https://postgrest.org/en/v12/references/api/tables_views.html#logical-operators
So probably `.or('col.not.is.null')` or worst case `.or('not.and(col.is.null)')`

**Reply by RafaelCENG** [reply]
Hi im trying something like this
```
currentBuilder
                       .not("athlete_bikes", "is", null)
                       .in("athlete_bikes.bike_types.name", values as Array<string>,)
```
I'm trying to do a filter where athlete_bikes is not null and then filter the name based on the array. My problem is the .not. Now that i can have athlete_bikes null column the problem arise.

I should clarify that athlete_bikes is an other table.

**Reply by serhanguney** [reply]
Just FYI - I was also having issues with using `not` with referenced tables. My issue seems to be fixed by querying the referenced table with inner join:
```
// So instead of
await supabase
  .from('instruments')
  .select('name, orchestral_sections(*)').not('orchestral_sections.column','is',null)

// you go with
await supabase
  .from('instruments')
  .select('name, orchestral_sections!inner(name)').not('orchestral_sections.column','is',null)

``` 

I hope it helps.

**Comment by Mihai-github** [comment]
Hello,

Is anyone getting timeout from on supabase when performing the is("column_name", null)?
